bulbs/reading_list/mixins.py

```diff
@@ -0,0 +1,209 @@
+from bulbs.content.models import Content
+from bulbs.content.search import randomize_es
+from bulbs.sections.models import Section
+from bulbs.special_coverage.models import SpecialCoverage
+from .popular import get_popular_ids, popular_content
+from .slicers import FirstSlotSlicer
+
+
+READING_LIST_CONFIG = getattr(settings, "READING_LIST_CONFIG", {})
+
+
+class ReadingListMixin(object):
+    """Mixin for Content-based objects to manage reading lists."""
+
+    def _get_reading_list_identifier(self):
+
+        # 1. Match content to sponsored Special Coverages
+        results = self.percolate_special_coverage(sponsored_only=True)
+        if results:
+            return results[0]
+
+        # 2."Popular" i.e., the content is one of the 25 most popular items.
+        popular_ids = get_popular_ids()
+        if popular_ids and self.id in popular_ids:
+            return "popular"
+
+        # 3. Any unsponsored special coverage reading list that contains this item.
+        results = self.percolate_special_coverage()
+        if results:
+            return results[0]
+
+        # 4. Any section that contains this item.
+        try:
+            results = Content.search_objects.client.percolate(
+                index=self.mapping.index,
+                doc_type=self.mapping.doc_type,
+                id=self.id,
+                body={"filter": es_filter.Prefix(_id="section").to_dict()}
+            )
+        except TransportError:
+            results = {"total": 0}
+
+        if results["total"] > 0:
+            for result in results["matches"]:
+                if not result["_id"].endswith("None"):
+                    return result["_id"]
+
+        return "recent"
+
+    def validate_query(self, query):
+        """Confirm query exists given common filters."""
+        if query is None:
+            return query
+        query = self.update_reading_list(query)
+        return query
+
+    def get_augment_query(self, augment_query=None):
+        """
+        Common rules for reading list augmentation hierarchy.
+
+        1. Sponsored Content.
+        2. Video Content.
+        """
+        augment_query = self.validate_query(augment_query)
+
+        # Given an invalid query, reach for a Sponsored query.
+        if not augment_query:
+            augment_query = self.validate_query(Content.search_objects.sponsored())
+
+        # Given an invalid Sponsored query, reach for a Video query.
+        if not augment_query:
+            augment_query = self.validate_query(Content.search_objects.evergreen_video())
+
+        return augment_query
+
+    def augment_reading_list(self, primary_query, augment_query=None, reverse_negate=False):
+        """Apply injected logic for slicing reading lists with additional content."""
+        primary_query = self.validate_query(primary_query)
+        augment_query = self.get_augment_query(augment_query=augment_query)
+
+        try:
+            if not augment_query:
```

Comment:
I think these 2 lines can be moved out of try/except block. to keep it as small as possible